crates/bevy_ecs/src/system/query.rs

```diff
@@ -250,12 +252,167 @@ use std::{any::TypeId, borrow::Borrow, fmt::Debug};
+    //! the `Query`)
+    //!
+    //! A notable thing about this is that for non-readonly `Query`'s the world borrows are an `&'lock mut ...` this
+    //! helps to explain why the lifetime annotations on `Query` methods do not return borrows tied to `'qworld` but instead `'_`
+    //! (see [#4243](https://github.com/bevyengine/bevy/pull/4243)). In our methods we either have `&Self` or `&mut Self` which
+    //! also means that inside of `Query::get (_mut?)` we have:
+    //!
+    //! `&'a (mut?) &'lock mut HashMap<ArchetypeComponentId, ReadOrWriteGuard<'world, Column>>`
+    //!
+    //! and borrow check would naturally prevent us from creating an `&'a (mut?) Column` out of this, which is what the unsound fn sig:
+    //!
+    //! `fn get_mut(&mut self, e: Entity) -> QueryItem<'qworld, Q>`
+    //!
+    //! would require doing.
+    //!
+    //! ---
+    //!
+    //! The [`QueryWorldBorrows`] abstraction exists to help prevent this kind of impl footgun by never giving out a
+    //! `&'qworld World` without taking ownership of `QueryWorldBorrows` (conceptually similar to having a `&'a mut &'b mut T`
+    //! and consuming the `&'a mut ..` to get a `&'a mut T`)
+    //!
+    //! This abstraction also exists to make all ways of gaining access to the `&World` that `Query` requires, unsafe. Since the
+    //! `&World` does not actually mean you can access the entire world immutably, a large amount of safe APIs on `World` would
+    //! be unsound to call which is a footgun for anyone implementing `Query`.
+    //!
+    //! In the future this type may be extended to _actually_ store a borrow of a hashmap when `debug_assertions` are enabled
+    //! so that we can check that all usage of the query is consistent with what access we told the scheduler we required.
+
+    use super::*;
+    use std::marker::PhantomData;
+
+    /// Modelling `Query`'s borrow on `World` as `&'qworld World` is not "correct" for many reasons.
+    /// This struct intends to hide the `&'qworld World` and expose an API that is more similar to what
+    /// the `&'qworld World` conceptually represents.
+    ///
+    /// This struct only implements [`Copy`] and [`Clone`] when `QF: ReadOnlyWorldQuery` to mimic the fact that
+    /// `ReadOnlyWorldQuery` borrows should act like `&'lock HashMap<..., ...>` wheras mutable world queries act
+    /// like `&'lock mut HashMap<..., ...>` (see module level docs).
+    pub struct QueryWorldBorrows<'world, QF: WorldQuery> {
+        world: &'world World,
+        // invariance because its probably possible to have two worldquery's where one is the subtype of another
+        // but each have different mutabilities. i.e. `dyn for<'a> Trait<'a>: WorldQuery` and `dyn Trait<'static>: WorldQuery`.
+        // probably not unsound since `Q` and `F` are already invariant on `Query` but this seems like a footgun and I don't care
+        // to try and reason about if this could cause trouble when covariant.
+        _p: PhantomData<*mut QF>,
```

Comment:
This probably made `Query` !Send and !Sync need to fix that (or atleast test this)